components/feature/awesomebar/src/test/java/mozilla/components/feature/awesomebar/provider/BookmarksStorageSuggestionProviderTest.kt

```diff
@@ -0,0 +1,194 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.awesomebar.provider
+
+import kotlinx.coroutines.runBlocking
+import mozilla.appservices.places.BookmarkRoot
+import mozilla.components.concept.storage.BookmarkInfo
+import mozilla.components.concept.storage.BookmarkNode
+import mozilla.components.concept.storage.BookmarkNodeType
+import mozilla.components.concept.storage.BookmarksStorage
+import mozilla.components.support.test.mock
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import java.util.UUID
+
+@RunWith(RobolectricTestRunner::class)
+class BookmarksStorageSuggestionProviderTest {
+    val bookmarks = testableBookmarksStorage()
+    private val newItem = BookmarkNode(
+        BookmarkNodeType.ITEM, "123", "456", null,
+        "Mozilla", "http://www.mozilla.org", null
+    )
+
+    @Test
+    fun `Provider returns empty list when text is empty`() = runBlocking {
+        val provider = BookmarksStorageSuggestionProvider(mock(), mock())
+
+        val suggestions = provider.onInputChanged("")
+        assertTrue(suggestions.isEmpty())
+    }
+
+    @Test
+    fun `Provider returns suggestions from configured bookmarks storage`() = runBlocking {
+        val provider = BookmarksStorageSuggestionProvider(bookmarks, mock())
+
+        bookmarks.addItem(BookmarkRoot.Mobile.id, newItem.url!!, newItem.title!!, null)
+
+        val suggestions = provider.onInputChanged("moz")
+        assertEquals(1, suggestions.size)
+        assertEquals("http://www.mozilla.org", suggestions[0].description)
+    }
+
+    @Test
+    fun `Provider does not return duplicate suggestions`() = runBlocking {
+        val provider = BookmarksStorageSuggestionProvider(bookmarks, mock())
+
+        for (i in 1..20) {
+            bookmarks.addItem(BookmarkRoot.Mobile.id, newItem.url!!, newItem.title!!, null)
+        }
+
+        val suggestions = provider.onInputChanged("moz")
+        assertEquals(1, suggestions.size)
+    }
+
+    @Test
+    fun `Provider limits number of returned unique suggestions`() = runBlocking {
+        val provider = BookmarksStorageSuggestionProvider(bookmarks, mock())
```

Comment:
Perhaps you can just add an InMemoryBookmarkStorage to `browser-storage-memory`? You already did some of that work here. Up to you :)